File: examples/showcase/github_fronting/custom-code-bundles/github_fronting_csharp/BackendAdapter.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using Anip.Service;

namespace {{ANIP_CSHARP_ROOT_NAMESPACE}};

public delegate Dictionary<string, object?> BackendAdapterHandler(
    Dictionary<string, object?> capability,
    Dictionary<string, object?> plan,
    Dictionary<string, object?> adapterInput,
    InvocationContext context);

public static class BackendAdapter
{
    public static BackendAdapterHandler Default => Execute;

    private static readonly HttpClient Http = new();

    private static Dictionary<string, object?> Execute(
        Dictionary<string, object?> capability,
        Dictionary<string, object?> plan,
        Dictionary<string, object?> parameters,
        InvocationContext context)
    {
        var unresolved = StringList(plan.GetValueOrDefault("unresolved_required_backend_inputs"));
        if (unresolved.Count > 0)
        {
            return Result(capability, plan, "backend_input_incomplete", new() { ["unresolved_required_backend_inputs"] = unresolved });
        }

        var token = Environment.GetEnvironmentVariable("GITHUB_TOKEN")?.Trim() ?? "";
        if (string.IsNullOrWhiteSpace(token))
        {
            return Result(capability, plan, "backend_error", new() { ["github_error"] = new Dictionary<string, object?> { ["error"] = "missing_github_token" } });
        }

        return Text(capability.GetValueOrDefault("capability_id")) switch
        {
            "github.repo.search_context" => SearchRepositoryContext(capability, plan, parameters, token),
            "github.issue.prepare" => PrepareOrCreateIssue(capability, plan, parameters, token, context),
            "github.pr.comment.prepare" => PreparePullRequestComment(capability, plan, parameters, token, context),
            "github.workflow.dispatch.request" => PrepareWorkflowDispatch(capability, plan, parameters, token, context),
            "github.release_notes.prepare" => PrepareReleaseNotes(capability, plan, parameters, token),
            _ => Result(capability, plan, "backend_execution_stub", new() { ["note"] = "No GitHub custom handler is registered for this capability." }),
        };
    }

    private static Dictionary<string, object?> SearchRepositoryContext(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, string token)
    {
        var owner = Text(parameters.GetValueOrDefault("owner"));
        var repo = Text(parameters.GetValueOrDefault("repo"));
        if (!RepoAllowed(owner, repo)) return Restricted(capability, plan, owner, repo);
        var limit = BoundedLimit(parameters.GetValueOrDefault("limit"), 20, 50);
        var query = $"repo:{owner}/{repo} {Text(parameters.GetValueOrDefault("query"))}".Trim();
        var payload = GitHub("GET", $"/search/issues?q={Uri.EscapeDataString(query)}&per_page={limit}", token, null);
        if (payload.ContainsKey("error")) return Result(capability, plan, "backend_error", new() { ["github_error"] = payload });
        var items = IssueItems(payload.GetValueOrDefault("items"), limit);
        return Result(capability, plan, "completed", new() { ["github_query"] = query, ["result"] = new Dictionary<string, object?> { ["items"] = items, ["count"] = items.Count, ["total_count"] = payload.GetValueOrDefault("total_count") } });
    }

    private static Dictionary<string, object?> PrepareOrCreateIssue(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, string token, InvocationContext context)
    {
        var repoContext = RepoMetadata(capability, plan, parameters, token);
        if (repoContext.Error is not null) return repoContext.Error;
        var body = new Dictionary<string, object?> { ["title"] = Text(parameters.GetValueOrDefault("title")), ["body"] = Text(parameters.GetValueOrDefault("body")) };
        var labels = StringList(parameters.GetValueOrDefault("labels"));
        var assignees = StringList(parameters.GetValueOrDefault("assignees"));
        if (labels.Count > 0) body["labels"] = labels;
        if (assignees.Count > 0) body["assignees"] = assignees;
        var preview = WritePreview(capability, plan, "issues.create", $"/repos/{repoContext.Owner}/{repoContext.Repo}/issues", body, repoContext.Payload!);
        if (!MutationEnabled(context)) return preview;
        var created = GitHub("POST", $"/repos/{Uri.EscapeDataString(repoContext.Owner)}/{Uri.EscapeDataString(repoContext.Repo)}/issues", token, body);
        if (created.ContainsKey("error"))
        {
            preview["execution_status"] = "backend_error";
            preview["github_error"] = created;
            return preview;
        }
        preview["execution_status"] = "completed";
        preview["approval_required"] = false;
        preview["mutation_performed"] = true;
        preview["created_issue"] = new Dictionary<string, object?> { ["number"] = created.GetValueOrDefault("number"), ["html_url"] = created.GetValueOrDefault("html_url"), ["state"] = created.GetValueOrDefault("state") };
        preview["note"] = "Created GitHub issue after the ANIP runtime validated and reserved an approval grant.";
        return preview;
    }

    private static Dictionary<string, object?> PreparePullRequestComment(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, string token, InvocationContext context)
    {
        var repoContext = RepoMetadata(capability, plan, parameters, token);
        if (repoContext.Error is not null) return repoContext.Error;
        var pullNumber = Text(parameters.GetValueOrDefault("pull_number"));
        var pull = GitHub("GET", $"/repos/{Uri.EscapeDataString(repoContext.Owner)}/{Uri.EscapeDataString(repoContext.Repo)}/pulls/{Uri.EscapeDataString(pullNumber)}", token, null);
        if (pull.ContainsKey("error")) return Result(capability, plan, "backend_error", new() { ["github_error"] = pull });
        var body = new Dictionary<string, object?> { ["body"] = $"[{FirstNonEmpty(Text(parameters.GetValueOrDefault("comment_purpose")), "triage_update")}] {Text(parameters.GetValueOrDefault("context"))}".Trim() };
        var preview = WritePreview(capability, plan, "issues.createComment", $"/repos/{repoContext.Owner}/{repoContext.Repo}/issues/{pullNumber}/comments", body, repoContext.Payload!);
        preview["pull_request"] = new Dictionary<string, object?> { ["number"] = pull.GetValueOrDefault("number"), ["title"] = pull.GetValueOrDefault("title"), ["state"] = pull.GetValueOrDefault("state") };
        if (!MutationEnabled(context)) return preview;
        var posted = GitHub("POST", $"/repos/{Uri.EscapeDataString(repoContext.Owner)}/{Uri.EscapeDataString(repoContext.Repo)}/issues/{Uri.EscapeDataString(pullNumber)}/comments", token, body);
        if (posted.ContainsKey("error"))
        {
            preview["execution_status"] = "backend_error";
            preview["github_error"] = posted;
            return preview;
        }
        preview["execution_status"] = "completed";
        preview["approval_required"] = false;
        preview["mutation_performed"] = true;
        preview["posted_comment"] = new Dictionary<string, object?> { ["id"] = posted.GetValueOrDefault("id"), ["html_url"] = posted.GetValueOrDefault("html_url") };
        return preview;
    }

    private static Dictionary<string, object?> PrepareWorkflowDispatch(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, string token, InvocationContext context)
    {
        var repoContext = RepoMetadata(capability, plan, parameters, token);
        if (repoContext.Error is not null) return repoContext.Error;
        var workflowId = Text(parameters.GetValueOrDefault("workflow_id"));
        var body = new Dictionary<string, object?> { ["ref"] = Text(parameters.GetValueOrDefault("ref")), ["inputs"] = ObjectMap(parameters.GetValueOrDefault("inputs")) };
        var preview = WritePreview(capability, plan, "actions.createWorkflowDispatch", $"/repos/{repoContext.Owner}/{repoContext.Repo}/actions/workflows/{workflowId}/dispatches", body, repoContext.Payload!);
        if (!MutationEnabled(context)) return preview;
        var dispatched = GitHub("POST", $"/repos/{Uri.EscapeDataString(repoContext.Owner)}/{Uri.EscapeDataString(repoContext.Repo)}/actions/workflows/{Uri.EscapeDataString(workflowId)}/dispatches", token, body);
        if (dispatched.ContainsKey("error"))
        {
            preview["execution_status"] = "backend_error";
            preview["github_error"] = dispatched;
            return preview;
        }
        preview["execution_status"] = "completed";
        preview["approval_required"] = false;
        preview["mutation_performed"] = true;
        preview["dispatched_workflow"] = new Dictionary<string, object?> { ["workflow_id"] = workflowId, ["ref"] = body["ref"] };
        return preview;
    }

    private static Dictionary<string, object?> PrepareReleaseNotes(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, string token)
    {
        var repoContext = RepoMetadata(capability, plan, parameters, token);
        if (repoContext.Error is not null) return repoContext.Error;
        var range = Text(parameters.GetValueOrDefault("range"));
        return Result(capability, plan, "completed", new()
        {
            ["result"] = new Dictionary<string, object?>
            {
                ["title"] = $"Release notes for {repoContext.Owner}/{repoContext.Repo} {range}",
                ["audience"] = FirstNonEmpty(Text(parameters.GetValueOrDefault("audience")), "internal"),
                ["repository"] = RepoSummary(repoContext.Payload!),
                ["range"] = range,
                ["sections"] = new[]
                {
                    new Dictionary<string, object?> { ["title"] = "Highlights", ["items"] = new[] { "Review bounded GitHub context before publishing release notes." } },
                    new Dictionary<string, object?> { ["title"] = "Governance", ["items"] = new[] { "This capability drafts content only and does not create a GitHub release." } },
                },
            },
            ["mutation_performed"] = false,
        });
    }

    public static Dictionary<string, object?> GitHub(string method, string path, string token, Dictionary<string, object?>? body)
    {
        using var request = new HttpRequestMessage(new HttpMethod(method), $"https://api.github.com{path}");
        request.Headers.Accept.ParseAdd("application/vnd.github+json");
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        request.Headers.UserAgent.ParseAdd("anip-github-fronting-showcase");
        request.Headers.TryAddWithoutValidation("X-GitHub-Api-Version", "2022-11-28");
        if (body is not null) request.Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");
        var response = Http.Send(request);
        var text = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
        var payload = string.IsNullOrWhiteSpace(text) ? new Dictionary<string, object?>() : JsonSerializer.Deserialize<Dictionary<string, object?>>(text)!;
        if (!response.IsSuccessStatusCode) return new Dictionary<string, object?> { ["error"] = "github_http_error", ["status"] = (int)response.StatusCode, ["detail"] = payload };
        return payload;
    }

    private static RepoContext RepoMetadata(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, string token)
    {
        var owner = Text(parameters.GetValueOrDefault("owner"));
        var repo = Text(parameters.GetValueOrDefault("repo"));
        if (!RepoAllowed(owner, repo)) return new RepoContext(owner, repo, null, Restricted(capability, plan, owner, repo));
        var payload = GitHub("GET", $"/repos/{Uri.EscapeDataString(owner)}/{Uri.EscapeDataString(repo)}", token, null);
        if (payload.ContainsKey("error")) return new RepoContext(owner, repo, null, Result(capability, plan, "backend_error", new() { ["github_error"] = payload }));
        return new RepoContext(owner, repo, payload, null);
    }

    private static Dictionary<string, object?> WritePreview(Dictionary<string, object?> capability, Dictionary<string, object?> plan, string action, string path, Dictionary<string, object?> body, Dictionary<string, object?> repoPayload) =>
        Result(capability, plan, "prepared", new()
        {
            ["approval_required"] = true,
            ["mutation_performed"] = false,
            ["github_action"] = action,
            ["github_metadata"] = RepoSummary(repoPayload),
            ["github_request"] = new Dictionary<string, object?> { ["method"] = "POST", ["path"] = path, ["body"] = body },
            ["note"] = "Prepared a GitHub request payload. No GitHub mutation was performed.",
        });

    private static Dictionary<string, object?> RepoSummary(Dictionary<string, object?> payload)
    {
        var owner = ObjectMap(payload.GetValueOrDefault("owner"));
        return new Dictionary<string, object?> { ["owner"] = owner.GetValueOrDefault("login"), ["repo"] = payload.GetValueOrDefault("name"), ["default_branch"] = payload.GetValueOrDefault("default_branch"), ["private"] = payload.GetValueOrDefault("private"), ["html_url"] = payload.GetValueOrDefault("html_url") };
    }

    private static Dictionary<string, object?> Result(Dictionary<string, object?> capability, Dictionary<string, object?> plan, string status, Dictionary<string, object?> extra)
    {
        var result = new Dictionary<string, object?> { ["execution_status"] = status, ["capability_id"] = capability.GetValueOrDefault("capability_id"), ["selected_backend"] = plan.GetValueOrDefault("selected_binding"), ["semantic_input"] = plan.GetValueOrDefault("semantic_input"), ["backend_input_contract"] = plan.GetValueOrDefault("backend_input_contract") };
        foreach (var item in extra) result[item.Key] = item.Value;
        return result;
    }

    private static Dictionary<string, object?> Restricted(Dictionary<string, object?> capability, Dictionary<string, object?> plan, string owner, string repo) =>
        Result(capability, plan, "restricted", new() { ["repository"] = new Dictionary<string, object?> { ["owner"] = owner, ["repo"] = repo }, ["reason"] = "GitHub repository is outside the configured ANIP repository policy." });

    private static bool RepoAllowed(string owner, string repo)
    {
        var key = $"{owner}/{repo}".ToLowerInvariant();
        var blocked = CsvEnv("ANIP_GITHUB_BLOCKED_REPOS");
        var allowed = CsvEnv("ANIP_GITHUB_ALLOWED_REPOS");
        return !blocked.Contains(key) && (allowed.Count == 0 || allowed.Contains(key));
    }

    private static bool MutationEnabled(InvocationContext context) =>
        Environment.GetEnvironmentVariable("ANIP_GITHUB_ALLOW_MUTATION") == "true" && !string.IsNullOrWhiteSpace(context?.ApprovalGrant);

    private static List<Dictionary<string, object?>> IssueItems(object? value, int limit)
    {
        var result = new List<Dictionary<string, object?>>();
        if (value is JsonElement element && element.ValueKind == JsonValueKind.Array)
        {
            foreach (var item in element.EnumerateArray())
            {
                if (result.Count >= limit) break;
                result.Add(new()
                {
                    ["number"] = Get(item, "number"),
                    ["title"] = Get(item, "title"),
                    ["state"] = Get(item, "state"),
                    ["html_url"] = Get(item, "html_url"),
                    ["kind"] = item.TryGetProperty("pull_request", out _) ? "pull_request" : "issue",
                });
            }
        }
        return result;
    }

    private static Dictionary<string, object?> ObjectMap(object? value)
    {
        if (value is Dictionary<string, object?> dict) return dict;
        if (value is JsonElement element && element.ValueKind == JsonValueKind.Object)
        {
            return JsonSerializer.Deserialize<Dictionary<string, object?>>(element.GetRawText()) ?? new();
        }
        return new();
    }

    private static string Get(JsonElement element, string key) => element.TryGetProperty(key, out var value) ? (value.ValueKind == JsonValueKind.String ? value.GetString() ?? "" : value.ToString()) : "";
    private static string FirstNonEmpty(params string[] values) => values.FirstOrDefault(value => !string.IsNullOrWhiteSpace(value)) ?? "";
    private static int BoundedLimit(object? value, int defaultValue, int maximum) => int.TryParse(Text(value), out var parsed) ? Math.Max(1, Math.Min(parsed, maximum)) : defaultValue;
    private static string Text(object? value) => value is null ? "" : value.ToString()!.Trim();
    private static List<string> CsvEnv(string name) => (Environment.GetEnvironmentVariable(name) ?? "").Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries).Select(item => item.ToLowerInvariant()).Distinct().ToList();
    private static List<string> StringList(object? value) => value is IEnumerable<object?> items ? items.Select(Text).Where(item => item.Length > 0).Distinct().ToList() : value is string text ? text.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries).Distinct().ToList() : [];

    private sealed record RepoContext(string Owner, string Repo, Dictionary<string, object?>? Payload, Dictionary<string, object?>? Error);
}

File: examples/showcase/github_fronting/custom-code-bundles/github_fronting_java/src/main/java/{{ANIP_JAVA_PACKAGE_PATH}}/BackendAdapter.java
package {{ANIP_JAVA_PACKAGE_NAME}};

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import dev.anip.service.InvocationContext;

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@FunctionalInterface
public interface BackendAdapter {

    Map<String, Object> execute(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> adapterInput, InvocationContext context);

    static BackendAdapter defaultAdapter() {
        return new GitHubBackendAdapter();
    }
}

final class GitHubBackendAdapter implements BackendAdapter {
    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final HttpClient HTTP = HttpClient.newHttpClient();

    @Override
    public Map<String, Object> execute(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, InvocationContext context) {
        List<String> unresolved = stringList(plan.get("unresolved_required_backend_inputs"));
        if (!unresolved.isEmpty()) {
            return result(capability, plan, "backend_input_incomplete", record("unresolved_required_backend_inputs", unresolved));
        }
        String token = System.getenv().getOrDefault("GITHUB_TOKEN", "").trim();
        if (token.isBlank()) return result(capability, plan, "backend_error", record("github_error", record("error", "missing_github_token")));
        return switch (text(capability.get("capability_id"))) {
            case "github.repo.search_context" -> searchRepositoryContext(capability, plan, params, token);
            case "github.issue.prepare" -> prepareOrCreateIssue(capability, plan, params, token, context);
            case "github.pr.comment.prepare" -> preparePullRequestComment(capability, plan, params, token, context);
            case "github.workflow.dispatch.request" -> prepareWorkflowDispatch(capability, plan, params, token, context);
            case "github.release_notes.prepare" -> prepareReleaseNotes(capability, plan, params, token);
            default -> result(capability, plan, "backend_execution_stub", record("note", "No GitHub custom handler is registered for this capability."));
        };
    }

    private static Map<String, Object> searchRepositoryContext(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String token) {
        String owner = text(params.get("owner"));
        String repo = text(params.get("repo"));
        if (!repoAllowed(owner, repo)) return restricted(capability, plan, owner, repo);
        int limit = boundedLimit(params.get("limit"), 20, 50);
        String query = ("repo:" + owner + "/" + repo + " " + text(params.get("query"))).trim();
        Map<String, Object> payload = github("GET", "/search/issues?q=" + url(query) + "&per_page=" + limit, token, null);
        if (payload.containsKey("error")) return result(capability, plan, "backend_error", record("github_error", payload));
        List<Map<String, Object>> items = new ArrayList<>();
        for (Map<String, Object> item : mapList(payload.get("items"))) {
            if (items.size() >= limit) break;
            items.add(record("number", item.get("number"), "title", item.get("title"), "state", item.get("state"), "html_url", item.get("html_url"), "kind", item.get("pull_request") == null ? "issue" : "pull_request"));
        }
        return result(capability, plan, "completed", record("github_query", query, "result", record("items", items, "count", items.size(), "total_count", payload.get("total_count"))));
    }

    private static Map<String, Object> prepareOrCreateIssue(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String token, InvocationContext context) {
        RepoContext repo = repoMetadata(capability, plan, params, token);
        if (repo.error != null) return repo.error;
        Map<String, Object> body = record("title", text(params.get("title")), "body", text(params.get("body")));
        List<String> labels = stringList(params.get("labels"));
        List<String> assignees = stringList(params.get("assignees"));
        if (!labels.isEmpty()) body.put("labels", labels);
        if (!assignees.isEmpty()) body.put("assignees", assignees);
        Map<String, Object> preview = writePreview(capability, plan, "issues.create", "/repos/" + repo.owner + "/" + repo.repo + "/issues", body, repo.payload);
        if (!mutationEnabled(context)) return preview;
        Map<String, Object> created = github("POST", "/repos/" + urlPath(repo.owner) + "/" + urlPath(repo.repo) + "/issues", token, body);
        if (created.containsKey("error")) {
            preview.put("execution_status", "backend_error");
            preview.put("github_error", created);
            return preview;
        }
        preview.put("execution_status", "completed");
        preview.put("approval_required", false);
        preview.put("mutation_performed", true);
        preview.put("created_issue", record("number", created.get("number"), "html_url", created.get("html_url"), "state", created.get("state")));
        preview.put("note", "Created GitHub issue after the ANIP runtime validated and reserved an approval grant.");
        return preview;
    }

    private static Map<String, Object> preparePullRequestComment(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String token, InvocationContext context) {
        RepoContext repo = repoMetadata(capability, plan, params, token);
        if (repo.error != null) return repo.error;
        String pullNumber = text(params.get("pull_number"));
        Map<String, Object> pull = github("GET", "/repos/" + urlPath(repo.owner) + "/" + urlPath(repo.repo) + "/pulls/" + urlPath(pullNumber), token, null);
        if (pull.containsKey("error")) return result(capability, plan, "backend_error", record("github_error", pull));
        Map<String, Object> body = record("body", ("[" + firstNonBlank(params.get("comment_purpose"), "triage_update") + "] " + text(params.get("context"))).trim());
        Map<String, Object> preview = writePreview(capability, plan, "issues.createComment", "/repos/" + repo.owner + "/" + repo.repo + "/issues/" + pullNumber + "/comments", body, repo.payload);
        preview.put("pull_request", record("number", pull.get("number"), "title", pull.get("title"), "state", pull.get("state")));
        if (!mutationEnabled(context)) return preview;
        Map<String, Object> posted = github("POST", "/repos/" + urlPath(repo.owner) + "/" + urlPath(repo.repo) + "/issues/" + urlPath(pullNumber) + "/comments", token, body);
        if (posted.containsKey("error")) {
            preview.put("execution_status", "backend_error");
            preview.put("github_error", posted);
            return preview;
        }
        preview.put("execution_status", "completed");
        preview.put("approval_required", false);
        preview.put("mutation_performed", true);
        preview.put("posted_comment", record("id", posted.get("id"), "html_url", posted.get("html_url")));
        return preview;
    }

    private static Map<String, Object> prepareWorkflowDispatch(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String token, InvocationContext context) {
        RepoContext repo = repoMetadata(capability, plan, params, token);
        if (repo.error != null) return repo.error;
        String workflowId = text(params.get("workflow_id"));
        Map<String, Object> body = record("ref", text(params.get("ref")), "inputs", objectMap(params.get("inputs")));
        Map<String, Object> preview = writePreview(capability, plan, "actions.createWorkflowDispatch", "/repos/" + repo.owner + "/" + repo.repo + "/actions/workflows/" + workflowId + "/dispatches", body, repo.payload);
        if (!mutationEnabled(context)) return preview;
        Map<String, Object> dispatched = github("POST", "/repos/" + urlPath(repo.owner) + "/" + urlPath(repo.repo) + "/actions/workflows/" + urlPath(workflowId) + "/dispatches", token, body);
        if (dispatched.containsKey("error")) {
            preview.put("execution_status", "backend_error");
            preview.put("github_error", dispatched);
            return preview;
        }
        preview.put("execution_status", "completed");
        preview.put("approval_required", false);
        preview.put("mutation_performed", true);
        preview.put("dispatched_workflow", record("workflow_id", workflowId, "ref", body.get("ref")));
        return preview;
    }

    private static Map<String, Object> prepareReleaseNotes(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String token) {
        RepoContext repo = repoMetadata(capability, plan, params, token);
        if (repo.error != null) return repo.error;
        String range = text(params.get("range"));
        return result(capability, plan, "completed", record(
            "result", record(
                "title", "Release notes for " + repo.owner + "/" + repo.repo + " " + range,
                "audience", firstNonBlank(params.get("audience"), "internal"),
                "repository", repoSummary(repo.payload),
                "range", range,
                "sections", List.of(
                    record("title", "Highlights", "items", List.of("Review bounded GitHub context before publishing release notes.")),
                    record("title", "Governance", "items", List.of("This capability drafts content only and does not create a GitHub release."))
                )
            ),
            "mutation_performed", false
        ));
    }

    private static RepoContext repoMetadata(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String token) {
        String owner = text(params.get("owner"));
        String repo = text(params.get("repo"));
        if (!repoAllowed(owner, repo)) return new RepoContext(owner, repo, null, restricted(capability, plan, owner, repo));
        Map<String, Object> payload = github("GET", "/repos/" + urlPath(owner) + "/" + urlPath(repo), token, null);
        if (payload.containsKey("error")) return new RepoContext(owner, repo, null, result(capability, plan, "backend_error", record("github_error", payload)));
        return new RepoContext(owner, repo, payload, null);
    }

    static Map<String, Object> github(String method, String path, String token, Map<String, Object> body) {
        try {
            HttpRequest.Builder builder = HttpRequest.newBuilder(URI.create("https://api.github.com" + path))
                .header("Accept", "application/vnd.github+json")
                .header("Authorization", "Bearer " + token)
                .header("Content-Type", "application/json")
                .header("User-Agent", "anip-github-fronting-showcase")
                .header("X-GitHub-Api-Version", "2022-11-28");
            if ("POST".equals(method)) builder.POST(HttpRequest.BodyPublishers.ofString(MAPPER.writeValueAsString(body == null ? Map.of() : body)));
            else builder.GET();
            HttpResponse<String> response = HTTP.send(builder.build(), HttpResponse.BodyHandlers.ofString());
            Map<String, Object> payload = response.body().isBlank() ? new LinkedHashMap<>() : MAPPER.readValue(response.body(), new TypeReference<>() {});
            if (response.statusCode() < 200 || response.statusCode() >= 300) return record("error", "github_http_error", "status", response.statusCode(), "detail", payload);
            return payload;
        } catch (Exception exc) {
            return record("error", "github_http_error", "detail", exc.toString());
        }
    }

    private static Map<String, Object> writePreview(Map<String, Object> capability, Map<String, Object> plan, String action, String path, Map<String, Object> body, Map<String, Object> repoPayload) {
        return result(capability, plan, "prepared", record(
            "approval_required", true,
            "mutation_performed", false,
            "github_action", action,
            "github_metadata", repoSummary(repoPayload),
            "github_request", record("method", "POST", "path", path, "body", body),
            "note", "Prepared a GitHub request payload. No GitHub mutation was performed."
        ));
    }

    private static Map<String, Object> repoSummary(Map<String, Object> payload) {
        Map<String, Object> owner = objectMap(payload.get("owner"));
        return record("owner", owner.get("login"), "repo", payload.get("name"), "default_branch", payload.get("default_branch"), "private", payload.get("private"), "html_url", payload.get("html_url"));
    }

    private static Map<String, Object> result(Map<String, Object> capability, Map<String, Object> plan, String status, Map<String, Object> extra) {
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("execution_status", status);
        result.put("capability_id", capability.get("capability_id"));
        result.put("selected_backend", plan.get("selected_binding"));
        result.put("semantic_input", plan.get("semantic_input"));
        result.put("backend_input_contract", plan.get("backend_input_contract"));
        result.putAll(extra);
        return result;
    }

    private static Map<String, Object> restricted(Map<String, Object> capability, Map<String, Object> plan, String owner, String repo) {
        return result(capability, plan, "restricted", record("repository", record("owner", owner, "repo", repo), "reason", "GitHub repository is outside the configured ANIP repository policy."));
    }

    private static boolean repoAllowed(String owner, String repo) {
        String key = (owner + "/" + repo).toLowerCase();
        List<String> blocked = csvEnv("ANIP_GITHUB_BLOCKED_REPOS");
        List<String> allowed = csvEnv("ANIP_GITHUB_ALLOWED_REPOS");
        return !blocked.contains(key) && (allowed.isEmpty() || allowed.contains(key));
    }

    private static boolean mutationEnabled(InvocationContext context) {
        String grant = context == null ? "" : text(context.getApprovalGrant());
        return "true".equals(System.getenv().getOrDefault("ANIP_GITHUB_ALLOW_MUTATION", "")) && !grant.isBlank();
    }

    private static List<String> csvEnv(String name) {
        List<String> result = new ArrayList<>();
        for (String item : System.getenv().getOrDefault(name, "").split(",")) {
            String value = item.trim().toLowerCase();
            if (!value.isBlank() && !result.contains(value)) result.add(value);
        }
        return result;
    }

    private static int boundedLimit(Object value, int defaultValue, int max) {
        try {
            return Math.max(1, Math.min(Integer.parseInt(text(value)), max));
        } catch (Exception ignored) {
            return defaultValue;
        }
    }

    private static List<String> stringList(Object value) {
        List<String> result = new ArrayList<>();
        if (value instanceof List<?> items) {
            for (Object item : items) if (!text(item).isBlank()) result.add(text(item));
        } else if (value instanceof String raw) {
            for (String part : raw.split(",")) if (!part.trim().isBlank()) result.add(part.trim());
        }
        return result;
    }

    @SuppressWarnings("unchecked")
    private static List<Map<String, Object>> mapList(Object value) {
        List<Map<String, Object>> result = new ArrayList<>();
        if (value instanceof List<?> items) for (Object item : items) if (item instanceof Map<?, ?> map) result.add((Map<String, Object>) map);
        return result;
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> objectMap(Object value) {
        return value instanceof Map<?, ?> map ? (Map<String, Object>) map : new LinkedHashMap<>();
    }

    private static String firstNonBlank(Object... values) {
        for (Object value : values) if (!text(value).isBlank()) return text(value);
        return "";
    }

    private static String text(Object value) {
        return value == null ? "" : value.toString().trim();
    }

    private static String url(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    private static String urlPath(String value) {
        return url(value).replace("+", "%20");
    }

    private static Map<String, Object> record(Object... pairs) {
        Map<String, Object> result = new LinkedHashMap<>();
        for (int index = 0; index + 1 < pairs.length; index += 2) result.put(String.valueOf(pairs[index]), pairs[index + 1]);
        return result;
    }

    private record RepoContext(String owner, String repo, Map<String, Object> payload, Map<String, Object> error) {}
}
